python_modules/dagster/dagster/_core/definitions/assets_job.py

```diff
@@ -376,70 +375,75 @@ def build_node_deps(
             node_alias = node_name
 
         # unique handle for each AssetsDefinition
-        assets_defs_by_node_handle[NodeHandle(node_alias, parent=None)] = assets_def
-        for output_name, key in assets_def.keys_by_output_name.items():
-            node_alias_and_output_by_asset_key[key] = (node_alias, output_name)
+        node_handle_to_asset[NodeHandle(node_alias, parent=None)] = asset
+
+        # Observation outputs are not recorded because they aren't a node dependency
+        if isinstance(asset, AssetsDefinition) and asset.is_materializable:
+            for output_name, key in asset.keys_by_output_name.items():
+                node_alias_and_output_by_asset_key[key] = (node_alias, output_name)
 
     asset_checks_defs_by_node_handle: Dict[NodeHandle, AssetChecksDefinition] = {}
-    for asset_checks_def in asset_checks_defs:
+    for asset_checks_def in asset_checks_to_execute:
         node_def_name = asset_checks_def.node_def.name
         node_key = NodeInvocation(node_def_name)
         asset_checks_defs_by_node_handle[NodeHandle(node_def_name, parent=None)] = asset_checks_def
 
     blocking_asset_check_output_handles_by_asset_key = (
         _get_blocking_asset_check_output_handles_by_asset_key(
-            assets_defs_by_node_handle, asset_checks_defs_by_node_handle
+            node_handle_to_asset, asset_checks_defs_by_node_handle
         )
     )
 
     deps: Dict[NodeInvocation, Dict[str, IDependencyDefinition]] = {}
-    for node_handle, assets_def in assets_defs_by_node_handle.items():
+    for node_handle, asset in node_handle_to_asset.items():
         # the key that we'll use to reference the node inside this AssetsDefinition
-        node_def_name = assets_def.node_def.name
+        node_def_name = check.not_none(asset.node_def).name
         alias = node_handle.name if node_handle.name != node_def_name else None
         node_key = NodeInvocation(node_def_name, alias=alias)
         deps[node_key] = {}
 
         # connect each input of this AssetsDefinition to the proper upstream node
-        for input_name in assets_def.input_names:
-            upstream_asset_key = resolved_asset_deps.get_resolved_asset_key_for_input(
-                assets_def, input_name
-            )
+        if isinstance(asset, AssetsDefinition) and asset.is_materializable:
```

Comment:
consider an early continue if possible to avoid tons of indenting and diff thrash